itest/itest_lndharness_test.go

```diff
@@ -0,0 +1,297 @@
+	"google.golang.org/grpc/credentials"
+)
+
+// lndHarness manages a single lndsignerd-backed instance of LND.
+type lndHarness struct {
+	tctx     *testContext
+	idPubKey string
+
+	unixSocket bool
+
+	cancel context.CancelFunc
+
+	lndSignerCmd *exec.Cmd
+
+	lndDir    string
+	lncliPath string
+	rpc       string
+	p2p       string
+	lndCmd    *exec.Cmd
+
+	startChan chan struct{}
+}
+
+// Start takes the initial configuration (tctx, idPubKey, and unixSocket) and
+// starts lndsignerd and LND.
+func (l *lndHarness) Start() {
+	l.tctx.t.Helper()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	l.cancel = cancel
+
+	// Make a channel. We'll close this channel after the node is fully
+	// started to signal clients it's safe to make calls.
+	l.startChan = make(chan struct{})
+
+	// Start lndsignerd.
+	l.lndDir = path.Join(l.tctx.tmpRoot, fmt.Sprintf("lnd%s", l.idPubKey))
+	err := os.Mkdir(l.lndDir, fs.ModeDir|0700)
+	require.NoError(l.tctx.t, err)
+
+	signerAddr := "127.0.0.1:" + newPortString()
+	fullSignerAddr := "tcp://" + signerAddr
+
+	if l.unixSocket {
+		signerAddr = "unix://" + path.Join(l.tctx.tmpRoot, l.idPubKey+".socket")
+		fullSignerAddr = signerAddr
+	}
+
+	l.lndSignerCmd = exec.CommandContext(ctx, l.tctx.lndSignerPath,
+		"--rpclisten="+fullSignerAddr, "--nodepubkey="+l.idPubKey,
+		"--tlscertpath=./testdata/tls.cert",
+		"--tlskeypath=./testdata/tls.key", "--network=regtest",
+	)
+
+	l.lndSignerCmd.Env = append(l.lndSignerCmd.Env,
+		"VAULT_ADDR=http://127.0.0.1:"+l.tctx.vaultPort,
+		"VAULT_TOKEN=root",
+	)
+
+	go waitProc(l.lndSignerCmd)
+
+	// Start lnd.
+	acctsResp, err := l.tctx.vaultClient.ReadWithData(
+		"lndsigner/lnd-nodes/accounts",
+		map[string][]string{
+			"node": []string{l.idPubKey},
+		},
+	)
+	require.NoError(l.tctx.t, err)
+
+	acctList, ok := acctsResp.Data["acctList"].(string)
+	require.True(l.tctx.t, ok)
+
+	accounts, err := lndsigner.GetAccounts(acctList)
+	require.NoError(l.tctx.t, err)
+
+	grpcAccounts := make([]*itest.WatchOnlyAccount, 0,
+		len(accounts))
+
+	for derPath, xPub := range accounts {
+		grpcAccounts = append(grpcAccounts,
+			&itest.WatchOnlyAccount{
+				Purpose:  derPath[0],
+				CoinType: derPath[1],
+				Account:  derPath[2],
+				Xpub:     xPub,
+			})
+	}
+
+	l.rpc = newPortString()
+	l.p2p = newPortString()
+
+	l.lndCmd = exec.CommandContext(ctx, l.tctx.lndPath,
+		"--lnddir="+l.lndDir, "--norest", "--listen="+l.p2p,
+		"--rpclisten="+l.rpc, "--trickledelay=1", "--bitcoin.active",
+		"--bitcoin.regtest", "--bitcoin.node=bitcoind",
+		"--bitcoind.rpcuser=user", "--bitcoind.rpcpass=password",
+		"--bitcoind.rpchost=127.0.0.1:"+l.tctx.bitcoinRPC,
+		"--bitcoind.zmqpubrawblock=tcp://"+l.tctx.bitcoinZB.String(),
+		"--bitcoind.zmqpubrawtx=tcp://"+l.tctx.bitcoinZT.String(),
+		"--remotesigner.enable",
+		"--remotesigner.rpchost="+signerAddr,
+		"--remotesigner.tlscertpath=./testdata/tls.cert",
+		"--remotesigner.macaroonpath=./testdata/signer.custom.macaroon",
```

Comment:
nitpick: we could generate TLS and macroons dynamically

Reply:
Re. the macaroon, it's possible because the macaroon is just a dummy file that's not checked by `lndsigner`. The `lnd` signer client only checks that its structure is valid before passing to `lndsigner`, not the actual auth data inside. This same macaroon can be used in production for that reason. We can generate macaroons dynamically, but that would involve adding macaroon bakery dependencies to the repo just for the itests. Let me know if you still want it.

Re. the TLS key/cert, I can add dynamic generation.

Reply:
I've added this in https://github.com/bottlepay/lndsigner/pull/11/commits/e2b734c6f3707941ef6f6c82fb094d22b937bcaa. If you're good with this, I'll squash it into the previous commit (adding integration tests). Now I'm going to look at generating a dummy macaroon without using the macaroon libraries.

Reply:
OK, I've also added dynamic macaroon generation using just manual byte slice ops in https://github.com/bottlepay/lndsigner/pull/11/commits/b9af6b2f5a2b5e6d2078f45941a3a564fbbb7212. If you're good with this, I'll squash it into the main integration test commit as well.

Reply:
LGTM